lib/ace/transport_app.rb

```diff
@@ -33,6 +34,44 @@ def initialize(config = nil)
       super(nil)
     end
 
+    def self.init_puppet_target(certname, transport, target)
+      unless target
+        ACE::Error.new("There was an error parsing the Puppet target. 'target' not found",
+                       'ace/puppet-target-error')
+        return nil
+      end
+      unless certname
+        ACE::Error.new("There was an error parsing the Puppet compiler details. 'certname' not found",
+                       'ace/puppet-target-error')
+        return nil
+      end
+      unless transport
+        ACE::Error.new("There was an error parsing the Puppet target. 'transport' not found",
+                       'ace/puppet-target-error')
+        return nil
+      end
+
+      if target['uri']
+        if target['uri'] =~ URI::DEFAULT_PARSER.make_regexp
+          # Correct URL
+          url = target['uri']
+        else
+          ACE::Error.new("There was an error parsing the URI of the Puppet target",
+                         'ace/puppet-target-error')
+          return nil
+        end
+      else
+        url = target
+      end
+
+      device_struct = Struct.new(:provider, :url, :name, :options)
+      device = Puppet::Util::NetworkDevice.init(device_struct.new(transport,
+                                                                  url,
```

Comment:
One thing discovered is that the keys on the hash need to be symbols for the RSAPI validation, so maybe something like: `Hash[url.map{|(k,v)| [k.to_sym,v]}]` but then for f5 uri's this could be problematic

Reply:
Thanks Thomas. Further up we decide if we are using the uri string or the hash, so I've added the Hash symbol conversion there.